README
README for the Auto Nice Daemon, AND
------------------------------------

The auto nice daemon will renice or even kill jobs according to a priority
database, after they take up too much CPU time. (You define what "too much"
actually means.) Refer to the man pages, and(8), and.conf(5), and
and.priorities(5), for details and instructions, and check the home page,
http://and.sourceforge.net/.


Platforms:

   Digital UNIX 4.0, 5.1
   FreeBSD 4.x
   IRIX and IRIX64
   Linux 2.2.x, 2.4.x and 2.6.x
   OpenBSD 2.7+
   Solaris 5.6

Requires:

   GNU make
   ANSI C Compiler

Documentation and Download:

   http://and.sourceforge.net/

Author:

   Patrick Schemitz <schemitz@users.sourceforge.net>

Credits:

   SRPM spec file help by Terje Rosten <terjeros@phys.ntnu.no>

   Linux/AXP jiffies to seconds fix by Markus Lischka
   <Markus_Lischka@physik.tu-muenchen.de>

   Debian package, Debian init script, Debian Makefile patches
   by Andras Bali <bali@debian.org>

   Debian Makefile patch by Mikael Andersson <mikan@mikan.net>

   The guys at INTEXXIA, http://www.intexxia.com, noticed a
   format string vulnerability and provided me with a patch.

   Pauli K. Borodulin <boro@fixel.net> pointed out that overwriting
   existing config files when doing make install was rude. He is
   right.

   Janet Casey <jcasey@gnu.org> noticed and reported that the LICENSE
   file was truncated.

   Dan Stromberg <strombrg@nis.acs.uci.edu> pointed out that
   the Digital UNIX version, and-OSF1.c, works virtually unchanged
   on IRIX, IRIX64 and Solaris (SunOS).

   "Quake2" <quake2@vladimir.eso.nu> pointed out that
   the OpenBSD version works virtually unchanged on FreeBSD.

   Marcelo Matus <mmatus@dinha.acms.arizona.edu> sent a patch for
   problems with very long running processes (alread longer than 248
   days when auto nice daemon is started), and other issues.

   Dr. Hans Ekkehard Plesser <hans.ekkehard.plesser@nlh.no> came up with
   the idea of also examining a process' parent, as well as with the
   minuid and mingid configuration options.

   Both xavier@rootshell.be and Jerome Warnier <jwarnier@beeznest.net>
   noticed that the Linux version accounted for usr time only; Xavier
   also provided a (one-line) fix. (Solaris, IRIX, and Tru 64 already
   did this.)

   Jerome and Martin Braure de Calignon <braurede@free.fr> noted and
   fixed a minor bug in the and.conf man page (default v. defaultnice).

   Jan Starzynski pointed out that under some Linux variants,
   /proc/<pid>/stat is always owned by root and thus unsuitable for
   obtaining the job's owner/group. He also provided a one-line fix.

Installation:

   Edit the Makefile, which is well documented. (g)make. (g)make install.
   (make simpleinstall if you don't have install(1) (which you really
   should). Edit the configuration files, /etc/and/and.conf and
   /etc/and/and.priorities. Start /usr/sbin/and. You must run it as root
   if you want it to renice or kill any jobs but your own; on all platforms
   but Linux, not even dummy mode will work for mortal users. That's due to
   the way process information is accessible under these Unices. (Linux is
   more generous here, which can be seen as both an advantage and a security
   flaw. I'm not conclusive on this topic.)

Last updated:

   This document was last updated 2005/03/27 by Patrick Schemitz
   <schemitz@users.sourceforge.net>


CHANGELOG
Changelog for Auto Nice Daemon
------------------------------

Changes in 1.2.2
- under some Linux variants, /proc/<pid>/stat always belongs to root
  so use /proc/<pid> to determine job owner/group
  (thanks to Jan Starzynski)
- fixed minor bug in manpage (defaultnice instead of default)
  (thanks to Martin Braure de Calignon and Jerome Warnier)

Changes in 1.2.1
- take into account not only usr but also sys time (Linux only; other
  Unixes already did this)

Changes in 1.2.0
- more robust signal handling/config reload
- startup script fixed for Tru64 killall
- proper daemon mode
- improved 64bit hardness
- verbosity handling now in and_printf()
- additional criterion: parent process (parent=.../ancestor=...)
  (thanks to Dr. Hans Ekkehard Plesser, NLH, for idea and testing)
- additional criterion: minimum user/group id
  (again thanks to Dr. Hans Ekkehard Plesser)

Changes in 1.0.9
- fix broken zombie fix (Linux only)
- fix ultra-long running (248+ days) processes (thanks to Marcelo
  Matus, UA)
- clean up signed/unsigned time vars
- compiled for AMD Opteron

Changes in 1.0.8
- leave zombies alone (Linux only)

Changes in 1.0.7
- port to FreeBSD (thanks to a guy who calls himself quake2)
- version and date are now compiled in
- make doc target to bump version number in man pages

Changes in 1.0.6
- stop overwriting existing configuration when make install
- small docs improvement (command name and regexes)
- changed misleading log message (seemingly negative nice level)
- added gtop to default config file
- fix LICENSE (still GPL, but the file was truncated)
- port to IRIX, IRIX64 and SunOS (thanks to Dan Stromberg, UCI)

Changes in 1.0.5
- fix format string vulnerability (please UPDATE!)
  (thanks to INTEXXIA.com)

Changes in 1.0.4
- makefile improvements, esp. for Debian (thanks to Andras Bali
  and Mikael Andersson)
- include LICENSE

Changes in 1.0.3
- bug fix Linux/AXP: jiffies to seconds fixed (thanks to Markus
  Lischka, TUM)
- in config files, "on" hostname is now an extended regex
  (thanks to Markus Lischka, TUM)
- new commandline switch -s for output to stdout
- documentation updates

Changes in 1.0.2
- minor documentation updates
- improved default configuration
- distribution-independend spec file for source rpms (thanks to
  Terje Rosten for his help)
- improved build process to support distribution-independend
  spec file
- fix logging: in -x mode (default), log via syslog; in -t mode (test),
  log to ./debug.and
- added chkconfig hook in SysV init script

Changes in 1.0.1
- fix status check in SysV init script

Changes in 1.0.0
- first official release


and.h
/*

    AND auto nice daemon - renice programs according to their CPU usage.
    Copyright (C) 1999-2004 Patrick Schemitz <schemitz@users.sourceforge.net>
    http://and.sourceforge.net/

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

#ifndef AND_H
#define AND_H

#include <stdbool.h>

/************************************************************************
 *                                                                      *
 * and.h -- AND library for platform-independent code.                  *
 *                                                                      *
 * 1999, 2004 Patrick Schemitz <schemitz@users.sourceforge.net>         *
 * http://and.sourceforge.net/                                          *
 *                                                                      *
 ***********************************************************************/

/*
 * and_procent -- process entry.
 *
 * AND-relevant information on a process.
 */
struct and_procent {
  /* to be filled by and-$OS.c: */
  int pid;
  int ppid;
  int uid;
  int gid;
  int nice;
  unsigned utime;
  char command [1024];
  /* to be filled by and.c: */
  struct and_procent *parent;
  struct and_procent *next;
};

/*
 * and_printf() - log message.
 *
 * Logs a message (in printf() format), either to stderr, or to syslog(),
 * depending on AND operational mode. In test mode (and -t), stderr is
 * used; syslog() otherwise. Use this to report any O/S-specific problems.
 * Log is suppressed if the current verbosity level is below the required
 * one.
 */
void and_printf (int required_verbosity, char *fmt, ...);

/*
 * and_setprocreader() -- set O/S specific handler for reading processes.
 *
 * getfirst and getnext are two functions returning a pointer to an
 * and_procent, or NULL if no more processes are available. The implementation
 * of these two functions are O/S specific. For Linux, reading through the
 * /proc filesystem is most suitable. See and-linux.c for a sample
 * implementation.
 *
 * Note: it is getfirst's task to also clean up any remainders of former
 * calls to getfirst and getnext, such as open DIR*.
 */
void and_setprocreader (struct and_procent *(*getfirst)(),
			struct and_procent *(*getnext)(),
			struct and_procent *(*getfrompid)());

/*
 * and_main() -- start the AND.
 *
 * Takes over control. Call this after setting the and_procreader().
 */
int and_main (int argc, char** argv);

#endif


and-proc.c
#include <sys/socket.h>
#include <linux/netlink.h>
#include <linux/connector.h>
#include <linux/cn_proc.h>
#include <signal.h>
#include <errno.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

#include "and.h"

bool need_exit = false;
/*
 * connect to netlink
 * returns netlink socket, or -1 on error
 */
int nl_connect()
{
	int rc;
	int nl_sock;
	struct sockaddr_nl sa_nl;

	nl_sock = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_CONNECTOR);
	if (nl_sock == -1) {
		perror("socket");
		return -1;
	}

	sa_nl.nl_family = AF_NETLINK;
	sa_nl.nl_groups = CN_IDX_PROC;
	sa_nl.nl_pid = getpid();

	rc = bind(nl_sock, (struct sockaddr *)&sa_nl, sizeof(sa_nl));
	if (rc == -1) {
		perror("bind");
		close(nl_sock);
		return -1;
	}

	return nl_sock;
}

/*
 * subscribe on proc events (process notifications)
 */
int proc_set_ev_listen(int nl_sock, bool enable)
{
	int rc;
	struct __attribute__ ((aligned(NLMSG_ALIGNTO))) {
		struct nlmsghdr nl_hdr;
		struct __attribute__ ((__packed__)) {
			struct cn_msg cn_msg;
			enum proc_cn_mcast_op cn_mcast;
		};
	} nlcn_msg;

	memset(&nlcn_msg, 0, sizeof(nlcn_msg));
	nlcn_msg.nl_hdr.nlmsg_len = sizeof(nlcn_msg);
	nlcn_msg.nl_hdr.nlmsg_pid = getpid();
	nlcn_msg.nl_hdr.nlmsg_type = NLMSG_DONE;

	nlcn_msg.cn_msg.id.idx = CN_IDX_PROC;
	nlcn_msg.cn_msg.id.val = CN_VAL_PROC;
	nlcn_msg.cn_msg.len = sizeof(enum proc_cn_mcast_op);

	nlcn_msg.cn_mcast = enable ? PROC_CN_MCAST_LISTEN : PROC_CN_MCAST_IGNORE;

	rc = send(nl_sock, &nlcn_msg, sizeof(nlcn_msg), 0);
	if (rc == -1) {
		perror("netlink send");
		return -1;
	}

	return 0;
}

/*
 * handle a single process event
 */
int proc_handle_ev(int nl_sock)
{
	int rc;

	struct __attribute__ ((aligned(NLMSG_ALIGNTO))) {
		struct nlmsghdr nl_hdr;
		struct __attribute__ ((__packed__)) {
			struct cn_msg cn_msg;
			struct proc_event proc_ev;
		};
	} nlcn_msg;

	while (!need_exit) {
		rc = recv(nl_sock, &nlcn_msg, sizeof(nlcn_msg), 0);
		if (rc == 0) {
			/* shutdown? */
			return -1;
		} else if (rc == -1) {
			and_printf(0, "netlink receive error\n");
			return -1;
		}
		switch (nlcn_msg.proc_ev.what) {
			case PROC_EVENT_NONE:
				and_printf(1, "start listening to netlink...\n");
				break;
			case PROC_EVENT_FORK:
				//if (fork() == 0) /* handle event in child */
				//	return nlcn_msg.proc_ev.event_data.fork.parent_pid;
				//and_printf(1, "FORK pid=%d ppid=%d\n", nlcn_msg.proc_ev.event_data.fork.child_pid,nlcn_msg.proc_ev.event_data.fork.parent_pid);
				break;
			case PROC_EVENT_EXEC:
				if (fork() == 0) /* handle event in child */
					return nlcn_msg.proc_ev.event_data.exec.process_pid;
				and_printf(2, "EXEC pid=%d\n", nlcn_msg.proc_ev.event_data.exec.process_pid);
				break;
			case PROC_EVENT_UID:
			case PROC_EVENT_GID:
			case PROC_EVENT_SID:
			case PROC_EVENT_PTRACE:
			case PROC_EVENT_COMM:
			case PROC_EVENT_EXIT:
				break;
			default:
				and_printf(1, "unhandled proc_event %d\n", nlcn_msg.proc_ev.what);
				break;
		}
	}
	return 0;
}




and-Linux.c
/*
    AND auto nice daemon - renice programs according to their CPU usage.
    Copyright (C) 1999-2003 Patrick Schemitz <schemitz@users.sourceforge.net>
    http://and.sourceforge.net/

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <unistd.h>
#include <string.h>
#include <dirent.h>
#include <ctype.h>
#include <syslog.h>
#include <sys/stat.h>
#include <signal.h>
#include <asm/param.h> /* kernel interrupt frequency: HZ */

#include "and.h"

/*
    AND -- auto-nice daemon/Linux version.

    Linux-specific AND version. Makes excessive use of the Linux
    /proc filesystem and is not portable.

    1999-2003 Patrick Schemitz, <schemitz@users.sourceforge.net>
    http://and.sourceforge.net/
*/

static DIR *linux_procdir = 0;
static struct and_procent linux_proc;

/*
    Scan /proc/<pid>/stat file. Format described in proc(5).
    l1: pid comm state ppid
    l2: pgrp session tty tpgid
    l3: flags minflt cminflt majflt cmajflt
    l4: utime stime cutime cstime
    l5: counter priority
*/
int linux_readproc (char *fn)
{
	FILE* f;
	int i;
	long li;
	long unsigned u, ujf, sjf;
	char state;
	char buffer [1024];
	if (!(f = fopen(fn,"rt"))) return 0;
	fscanf(f,"%d %1023s %c %d",&(linux_proc.pid),buffer,&state,&(linux_proc.ppid));
	fscanf(f,"%d %d %d %d",&i,&i,&i,&i);
	fscanf(f,"%lu %lu %lu %lu %lu",&u,&u,&u,&u,&u);
	fscanf(f,"%lu %lu %ld %ld",&ujf,&sjf,&li,&li);
	fscanf(f,"%ld %d",&li,&(linux_proc.nice));
	i = feof(f);
	fclose(f);
	if (i) return 0;
	if (state == 'Z') return 0;			/* ignore zombies */
	ujf += sjf;					/* take into account both usr and sys time */
	ujf /= HZ;					/* convert jiffies to seconds */
	linux_proc.utime = ujf > INT_MAX ? INT_MAX: (int) ujf;
	buffer[strlen(buffer)-1] = 0;			/* remove () around command name */
	strncpy(linux_proc.command,&buffer[1],1023);
	linux_proc.command[1023] = 0;
	and_printf(3, "Linux: process %s pid: %d ppid: %d\n",
		   linux_proc.command, linux_proc.pid, linux_proc.ppid);
	return 1;
}

struct and_procent *linux_getnext ()
{
	char name [1024];
	struct dirent *entry;
	struct stat dirstat;
	if (!linux_procdir) return NULL;
	while ((entry = readdir(linux_procdir)) != NULL) { /* omit . .. apm bus... */
		if (isdigit(entry->d_name[0])) break;
	}
	if (!entry) return NULL;
	/* stat() /proc/<pid> directory to get uid/gid */
	snprintf(name, 1024, "/proc/%s",entry->d_name);
	if (stat(name,&dirstat)) return linux_getnext();
	/* read the job's stat "file" to get command, nice level, etc */
	snprintf(name, 1024, "/proc/%s/stat",entry->d_name);
	if (!linux_readproc(name)) return linux_getnext();
	linux_proc.uid = dirstat.st_uid;
	linux_proc.gid = dirstat.st_gid;
	return &linux_proc;
}

struct and_procent *linux_getfrompid (int pid)
{
	char name [1024];
	struct stat dirstat;

	/* stat() /proc/<pid> directory to get uid/gid */
	snprintf(name, 1024, "/proc/%d", pid);
	stat(name,&dirstat);
	/* read the job's stat "file" to get command, nice level, etc */
	snprintf(name, 1024, "/proc/%d/stat", pid);
	linux_readproc(name);
	linux_proc.uid = dirstat.st_uid;
	linux_proc.gid = dirstat.st_gid;

	return &linux_proc;
}

struct and_procent *linux_getfirst ()
{
	if (linux_procdir) {
		rewinddir(linux_procdir);
	} else {
		linux_procdir = opendir("/proc");
		if (!linux_procdir) {
			and_printf(0,"cannot open /proc, aborting.\n");
			abort();
		}
	}
	return linux_getnext();
}

int main (int argc, char** argv)
{
	and_setprocreader(&linux_getfirst, &linux_getnext, &linux_getfrompid);
	return and_main(argc,argv);
}
